File: backend/app/schemas/production.py
from pydantic import BaseModel
from datetime import date
from uuid import UUID


class ProductionRecordCreate(BaseModel):
    production_date: date
    days_on: int = 30
    oil_rate: float | None = None
    gas_rate: float | None = None
    water_rate: float | None = None
    cum_oil: float | None = None
    cum_gas: float | None = None
    cum_water: float | None = None
    gor: float | None = None
    water_cut: float | None = None
    boe: float | None = None
    tubing_pressure: float | None = None
    casing_pressure: float | None = None
    flowing_bhp: float | None = None
    choke_size: float | None = None
    hours_on: float | None = None


class ProductionBatchCreate(BaseModel):
    records: list[ProductionRecordCreate]


class ProductionRecordResponse(BaseModel):
    well_id: UUID
    production_date: date
    days_on: int
    oil_rate: float | None
    gas_rate: float | None
    water_rate: float | None
    cum_oil: float | None
    cum_gas: float | None
    cum_water: float | None
    gor: float | None
    water_cut: float | None
    boe: float | None
    tubing_pressure: float | None
    casing_pressure: float | None
    flowing_bhp: float | None
    choke_size: float | None
    hours_on: float | None
    data_source: str
    is_validated: bool

    model_config = {"from_attributes": True}


class ProductionStatistics(BaseModel):
    total_records: int
    first_production_date: date | None
    last_production_date: date | None
    peak_oil_rate: float | None
    peak_gas_rate: float | None
    current_oil_rate: float | None
    current_gas_rate: float | None
    cum_oil: float | None
    cum_gas: float | None
    cum_water: float | None
    avg_oil_rate: float | None
    avg_gas_rate: float | None
    avg_water_cut: float | None
    avg_gor: float | None

File: frontend/src/types/production.ts
export interface ProductionRecord {
  well_id: string;
  production_date: string;
  days_on: number;
  oil_rate: number | null;
  gas_rate: number | null;
  water_rate: number | null;
  cum_oil: number | null;
  cum_gas: number | null;
  cum_water: number | null;
  gor: number | null;
  water_cut: number | null;
  boe: number | null;
  tubing_pressure: number | null;
  casing_pressure: number | null;
  flowing_bhp: number | null;
  choke_size: number | null;
  hours_on: number | null;
  data_source: string;
  is_validated: boolean;
}

export interface ProductionRecordCreate {
  production_date: string;
  days_on?: number;
  oil_rate?: number;
  gas_rate?: number;
  water_rate?: number;
  cum_oil?: number;
  cum_gas?: number;
  cum_water?: number;
  gor?: number;
  water_cut?: number;
  boe?: number;
  tubing_pressure?: number;
  casing_pressure?: number;
  flowing_bhp?: number;
  choke_size?: number;
  hours_on?: number;
}

export interface ProductionBatchCreate {
  records: ProductionRecordCreate[];
}

export interface ProductionStatistics {
  total_records: number;
  first_production_date: string | null;
  last_production_date: string | null;
  peak_oil_rate: number | null;
  peak_gas_rate: number | null;
  current_oil_rate: number | null;
  current_gas_rate: number | null;
  cum_oil: number | null;
  cum_gas: number | null;
  cum_water: number | null;
  avg_oil_rate: number | null;
  avg_gas_rate: number | null;
  avg_water_cut: number | null;
  avg_gor: number | null;
}

export interface ProductionFilters {
  start_date?: string;
  end_date?: string;
  page?: number;
  per_page?: number;
}

export interface BulkProductionWellSummary {
  well_identifier: string;
  well_name: string | null;
  records_imported: number;
  records_deleted?: number;
  status: "success" | "not_found" | "error";
  error?: string;
}

export interface BulkProductionImportResult {
  file_name: string;
  file_type: string;
  total_wells_detected: number;
  matched_wells: number;
  unmatched_wells: number;
  total_records_imported: number;
  per_well_summary: BulkProductionWellSummary[];
  errors: Array<{ well_identifier: string; message: string }>;
}
